quint/src/effects/substitutions.ts

```diff
@@ -86,7 +87,14 @@ export function applySubstitution(subs: Substitutions, e: Effect): Either<ErrorT
     }
   }
 
-  return result.map(simplify)
+  return result.map(simplify).chain(r => {
+    if (!isEqual(r, e)) {
+      // Keep reaplying the substitutions until the effect is unchanged.
+      // Useful when substitutions have a transitive pattern [ a |-> b, b |-> c ]
+      return applySubstitution(subs, r)
```

Comment:
Just to make sure: substitutions can never be cyclic?

Also, if not: do we really need to compute a fixed point here, or can we sort the substitutions topologically and apply them all at once?

Reply:
> However, I don't see a clear benefit of doing that over this?

idk, I'm always for avoiding recursion if things can be achieved in a single call.
I guess here the overhead does not matter much, because the arrays are small, but even so, small delays add up and I'd avoid them, especially for compile-time checks.

Reply:
Makes sense. I'll keep this for now and improve it a performance-focused pass in the future. There are way more pressing performance problems and, then, I'll find a way to be very careful to not make regressions with the optmizations.